src/metabase/models/card.clj

```diff
@@ -306,27 +337,55 @@
                            :query-database        query-db-id
                            :field-filter-database field-db-id})))))))
 
-(defn- assert-valid-model
+(defn- ensure-type-and-dataset-are-aligned
+  "We're in the process of migrating from using `report_card.dataset` to `report_card.type`.
+  In the future we'll drop `dataset` and only use `type`. But for now we need to make sure that both keys are aligned
+  when dealing with cards.
+  - If both keys are present, we make sure `dataset` is true if `type` is `model` else false.
+    This will make a different when we have `metric` type since a boolean can't represent tri-state
+  - If only one key is present, we'll assoc the correct value for the other key."
+  [{:keys [type dataset] :as card}]
+  (cond
+   ;; if none of the 2 keys presents, do nothing
+   (and (nil? type) (nil? dataset))
+   card
+
+   ;; if both type and dataset presents, we prioritize type
+   (and (some? type) (some? dataset))
+   (assoc card :dataset (= type "model"))
+
+   ;; if only type presents, make sure dataset follows
+   (some? type)
+   (assoc card :dataset (= type "model"))
+
+   ;; if only dataset presents, make sure type follows
+   (some? dataset)
+   (assoc card :type (if dataset
+                       "model"
+                       "question"))))
```

Comment:
I agree this is not ideal, I have this because a lot of our tests depend on creating a card with only `:dataset true`.

I think it's safe to have this because of 2 reasons:
- this relies on the idea that where the difference between "question, model," and "metric" matters, we should update the code to use "type."
- this function is only used during card creation and update, both code paths will be updated to use `type` instead of `dataset` (as noted in the PR's description). In production, this code path should never be reached, we will only see this in tests.

And I added a warning in [d87b778](https://github.com/metabase/metabase/pull/38074/commits/d87b7788674ac3c3d11d12af96b7972cb18cbd06)